eZ/Publish/Core/Persistence/Legacy/Content/UrlWildcard/Handler.php

```diff
@@ -113,4 +116,107 @@ public function loadAll($offset = 0, $limit = -1)
             $this->gateway->loadUrlWildcardsData($offset, $limit)
         );
     }
+
+    /**
+     * Performs lookup for given URL.
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException if the url wild card was not found
+     *
+     * @param string $sourceUrl
+     *
+     * @return \eZ\Publish\SPI\Persistence\Content\UrlWildcard
+     */
+    public function translate(string $sourceUrl): UrlWildcard
+    {
+        $row = $this->gateway->loadUrlWildcardBySourceUrl($sourceUrl);
+
+        if (!empty($row)) {
+            return $this->mapper->extractUrlWildcardFromRow($row);
+        }
+
+        // can't find UrlWildcard by simple lookup, continue and try to translate
+
+        $rows = $this->gateway->loadUrlWildcardsData();
+        uasort($rows,
+            static function ($row1, $row2) {
+                return strlen($row2['source_url']) - strlen($row1['source_url']);
+            }
+        );
+
+        foreach ($rows as $row) {
+            if ($uri = $this->match($sourceUrl, $row)) {
+                $row['destination_url'] = $uri;
+
+                return $this->mapper->extractUrlWildcardFromRow($row);
+            }
+        }
+
+        throw new NotFoundException('URLWildcard', $sourceUrl);
+    }
+
+    /**
+     * Checks whether UrlWildcard with given source url exits.
+     *
+     * @param string $sourceUrl
+     *
+     * @return bool
+     */
+    public function exactSourceUrlExists(string $sourceUrl): bool
+    {
+        $row = $this->gateway->loadUrlWildcardBySourceUrl($sourceUrl);
+
+        return !empty($row);
+    }
+
+    /**
+     * Tests if the given url matches against the given url wildcard.
+     *
+     * if the wildcard matches on the given url this method will return a ready
+     * to use destination url, otherwise this method will return <b>NULL</b>.
+     *
+     * @param string $url
+     * @param array $wildcard
+     *
+     * @return string|null
+     */
+    private function match(string $url, array $wildcard): ?string
+    {
+        if (preg_match($this->compile($wildcard['source_url']), $url, $match)) {
+            return $this->substitute($wildcard['destination_url'], $match);
+        }
+
+        return null;
+    }
+
+    /**
+     * Compiles the given url pattern into a regular expression.
+     *
+     * @param string $sourceUrl
+     *
+     * @return string
+     */
+    private function compile(string $sourceUrl): string
+    {
+        return '(^' . str_replace('\\*', '(.*)', preg_quote($sourceUrl)) . '$)U';
+    }
+
+    /**
+     * Substitutes all placeholders ({\d}) in the given <b>$destinationUrl</b> with
+     * the values from the given <b>$values</b> array.
+     *
+     * @param string $destinationUrl
+     * @param array $values
+     *
+     * @return string
+     */
+    private function substitute(string $destinationUrl, array $values): string
+    {
+        preg_match_all(self::PLACEHOLDERS_REGEXP, $destinationUrl, $matches);
```

Comment:
if `preg_match_all` doesn't match anything it will return `FALSE` and the next statement will trigger `E_NOTICE` warning.
I'm not sure however what is the correct way to handle it. Return `$destinationUrl` anyway or throw RuntimeException. Given that you don't seem to expect that due to current usage, I'd say it should throw.

Reply:
It's not correct. `preg_match_all` for this particular regexp will **always** return a two-dimensional array (ref: https://3v4l.org/jHXIh), even though it is empty. It might be empty if your destination URL leads to another domain. Therefore, `E_NOTICE` will never be triggered.

Reply:
> It's not correct. `preg_match_all` for this particular regexp will **always** return a two-dimensional array (ref: https://3v4l.org/jHXIh), even though it is empty. It might be empty if your destination URL leads to another domain. Therefore, `E_NOTICE` will never be triggered.

I know. Right now. This is why I approved the PR instead of requesting changes ;) 

It's rather about the fact that this construct, w/o sanity check is not a good pattern, looking entirely at this single method. What if regexp changes? What if the change will result in `E_NOTICE` only in edge cases? If function documents that it can return false and next lines of code depend on the expected execution of that code, it's always good idea to check that. Even if at the time of writing it, it won't occur.

Given this is not that important right now, you can keep it as-is.